File: cpp/src/p_0015.cc
#include <cstdint>

#include "euler.h"

using vector_u64 = std::vector<uint64_t>;

uint64_t euler_0015(int size) {
    std::vector<vector_u64> grid(size + 1, vector_u64(size + 1, 0));

    // Initialize the top row and left column
    for (int i = 0; i <= size; ++i) {
        grid[i][0] = 1;
        grid[0][i] = 1;
    }

    // Fill the grid
    for (int i = 1; i <= size; ++i) {
        for (int j = 1; j <= size; ++j) {
            grid[i][j] = grid[i - 1][j] + grid[i][j - 1];
        }
    }

    return grid[size][size];
}

File: zig/src/p_0015.zig
const std = @import("std");

fn euler_0015(comptime size: i8) usize {
    var grid: [size + 1][size + 1]usize = undefined;

    // Initialize the top row and the left column
    for (0..(size + 1)) |i| {
        grid[i][0] = 1;
        grid[0][i] = 1;
    }

    // Fill the grid
    for (1..(size + 1)) |i| {
        for (1..(size + 1)) |j| {
            grid[i][j] = grid[i - 1][j] + grid[i][j - 1];
        }
    }
    return grid[size][size];
}

test "euler15" {
    try std.testing.expectEqual(6, euler_0015(2));
}

pub fn main() !void {
    std.debug.print("Problem 0015: In a 20 x 20 grid the number of available routes are: {d}\n", .{euler_0015(20)});
}
